src/Common/rename.cpp

```diff
@@ -0,0 +1,122 @@
+#include <Common/rename.h>
+#include <Common/Exception.h>
+#include <Poco/File.h>
+
+#if defined(linux) || defined(__linux) || defined(__linux__)
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <linux/fs.h>
+#include <sys/utsname.h>
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int ATOMIC_RENAME_FAIL;
+    extern const int SYSTEM_ERROR;
+    extern const int UNSUPPORTED_METHOD;
+    extern const int FILE_ALREADY_EXISTS;
+}
+
+static bool supportsRenameat2Impl()
+{
+#if defined(__NR_renameat2)
```

Comment:
Maybe just try to execute `syscall(SYS_renameat2, ...)` with some invalid arguments? If `errno` will contain `ENOSYS` than syscall doesn't exist.

Reply:
We cannot execute `syscall(SYS_renameat2, ...)` if `SYS_renameat2` or `__NR_renameat2` is not defined. We can explicitly use its number like `syscall(316, ...)`, but it looks worse. Also I'm afraid of using syscalls on Mac OS and FreeBSD this way, so I [include](https://github.com/ClickHouse/ClickHouse/pull/7512/files/b29bddac12fe2defa35e551b54f1940dc89ab2b5#diff-540b5357dfacf6d06b8750aab586e8efR5) system headers only for Linux